lib/td/command/workflow.rb

```diff
@@ -0,0 +1,424 @@
+          filename = nil
+          tar.each do |entry|
+            # Handle LongLink
+            if entry.full_name == TAR_LONGLINK
+              filename = entry.read.strip
+              next
+            end
+            filename ||= entry.full_name
+
+            # Strip path components
+            stripped = strip_components(filename, strip)
+            filename = nil
+            if stripped.empty?
+              next
+            end
+            dest = File.join destination, stripped
+
+            if entry.directory? || (entry.header.typeflag == '' && entry.full_name.end_with?('/'))
+              File.rm_rf dest if File.file? dest
+              FileUtils.mkdir_p dest, :mode => entry.header.mode, :verbose => false
+            elsif entry.file? || (entry.header.typeflag == '' && !entry.full_name.end_with?('/'))
+              FileUtils.rm_rf dest if File.exist? dest
+              FileUtils.mkdir_p File.dirname dest
+              File.open dest, "wb" do |f|
+                f.print entry.read
+              end
+              FileUtils.chmod entry.header.mode, dest, :verbose => false
+            elsif entry.header.typeflag == '2' #Symlink!
+              File.symlink entry.header.linkname, dest
+            else
+              raise "Unkown tar entry: #{entry.full_name} type: #{entry.header.typeflag}."
+            end
+          end
+        end
+      end
+    end
+
+    def strip_components(filename, strip)
+      File.join(Pathname.new(filename).each_filename.drop(strip))
+    end
+
+    private
+    def check_digdag_cli
+      check_system_java unless bundled_java?
+
+      unless File.exists?(digdag_cli_path)
+        $stderr.puts 'td workflow module not yet installed, download now? [Y/n]'
+        line = $stdin.gets
+        line.strip!
+        if (not line.empty?) and (line !~ /^y(?:es)?$/i)
+          raise WorkflowError, 'Aborted'
+        end
+        download_digdag
+      end
+    end
+
+    def download_digdag
+      require 'net/http'
+      require 'openssl'
+
+      FileUtils.mkdir_p digdag_dir
+
+      if bundled_java?
+        download_java
+      end
+
+      Dir.mktmpdir do |download_dir|
+        indicator = Command::TimeBasedDownloadProgressIndicator.new(
+            'Downloading workflow module...', Time.new.to_i, 2)
+        status = nil
+        download_path = File.join(download_dir, 'digdag')
+        File.open(download_path, 'wb') do |file|
+          status = Updater.stream_fetch(digdag_url, file) {
+            indicator.update
+          }
+        end
+        indicator.finish
+
+        $stdout.puts
+
+        unless status
+          raise WorkflowError, 'Failed to download workflow module.'
+        end
+
+        $stdout.print 'Installing workflow module... '
+        FileUtils.rm_rf(digdag_cli_tmp_path)
+        FileUtils.cp(download_path, digdag_cli_tmp_path)
+        FileUtils.chmod('a=xr', digdag_cli_tmp_path)
+        FileUtils.mv(digdag_cli_tmp_path, digdag_cli_path)
```

Comment:
`digdag_cli_tmp_path` is in the the same directory as `digdag_cli_path` so the `mv` will be atomic. `download_path` might be on a different volume.